File: aib/app/services/session.py
"""
Session & message CRUD — with in-memory fallback when PostgreSQL is unavailable.
"""

from __future__ import annotations

import json
import uuid
from datetime import datetime, timezone

from app.database import query, query_one, is_db_available

# ═══════════════════════════════════════════════════════════════════════
# In-memory store (used when PostgreSQL is not available)
# ═══════════════════════════════════════════════════════════════════════
_sessions: dict[str, dict] = {}
_messages: dict[str, list[dict]] = {}  # session_id -> list of messages
_message_counter = 0


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _next_msg_id() -> int:
    global _message_counter
    _message_counter += 1
    return _message_counter


# ═══════════════════════════════════════════════════════════════════════
# Sessions
# ═══════════════════════════════════════════════════════════════════════

def create_session() -> dict:
    """Create a new active session and return it."""
    if is_db_available():
        return query_one(
            "INSERT INTO sessions (status) VALUES ('active') RETURNING *"
        )

    # In-memory fallback
    sid = str(uuid.uuid4())
    session = {
        "id": sid,
        "status": "active",
        "created_at": _now(),
        "updated_at": _now(),
        "completed_at": None,
    }
    _sessions[sid] = session
    _messages[sid] = []
    return session


def get_session(session_id: str) -> dict | None:
    """Get a session by UUID."""
    if is_db_available():
        return query_one("SELECT * FROM sessions WHERE id = %s", (session_id,))

    return _sessions.get(session_id)


def get_session_with_messages(session_id: str) -> dict | None:
    """Get a session together with its ordered messages."""
    if is_db_available():
        session = get_session(session_id)
        if session is None:
            return None
        messages = query(
            """SELECT id, role, content, attachments, created_at
               FROM messages
               WHERE session_id = %s
               ORDER BY created_at ASC""",
            (session_id,),
        )
        session["messages"] = messages
        return session

    # In-memory fallback
    session = _sessions.get(session_id)
    if session is None:
        return None
    result = dict(session)
    result["messages"] = list(_messages.get(session_id, []))
    return result


def list_sessions(status: str | None = None) -> list[dict]:
    """List sessions, optionally filtered by status."""
    if is_db_available():
        if status:
            return query(
                "SELECT * FROM sessions WHERE status = %s ORDER BY created_at DESC",
                (status,),
            )
        return query("SELECT * FROM sessions ORDER BY created_at DESC")

    # In-memory fallback
    sessions = list(_sessions.values())
    if status:
        sessions = [s for s in sessions if s["status"] == status]
    sessions.sort(key=lambda s: s["created_at"], reverse=True)
    return sessions


def update_session_status(session_id: str, status: str) -> dict | None:
    """Update a session's status (active → completed / abandoned)."""
    if is_db_available():
        completed_at = "NOW()" if status == "completed" else "NULL"
        return query_one(
            f"""UPDATE sessions
                SET status = %s, updated_at = NOW(), completed_at = {completed_at}
                WHERE id = %s
                RETURNING *""",
            (status, session_id),
        )

    # In-memory fallback
    session = _sessions.get(session_id)
    if session is None:
        return None
    session["status"] = status
    session["updated_at"] = _now()
    if status == "completed":
        session["completed_at"] = _now()
    return session


# ═══════════════════════════════════════════════════════════════════════
# Messages
# ═══════════════════════════════════════════════════════════════════════

def add_message(
    session_id: str,
    role: str,
    content: str,
    attachments: list | None = None,
) -> dict:
    """Insert a message and bump the session's updated_at timestamp."""
    attachments = attachments or []

    if is_db_available():
        msg = query_one(
            """INSERT INTO messages (session_id, role, content, attachments)
               VALUES (%s, %s, %s, %s)
               RETURNING *""",
            (session_id, role, content, json.dumps(attachments)),
        )
        query("UPDATE sessions SET updated_at = NOW() WHERE id = %s", (session_id,))
        return msg

    # In-memory fallback
    msg = {
        "id": _next_msg_id(),
        "session_id": session_id,
        "role": role,
        "content": content,
        "attachments": attachments,
        "created_at": _now(),
    }
    if session_id not in _messages:
        _messages[session_id] = []
    _messages[session_id].append(msg)

    # Bump session updated_at
    if session_id in _sessions:
        _sessions[session_id]["updated_at"] = _now()

    return msg


def get_messages(session_id: str) -> list[dict]:
    """Return all messages for a session, ordered chronologically."""
    if is_db_available():
        return query(
            """SELECT id, role, content, attachments, created_at
               FROM messages
               WHERE session_id = %s
               ORDER BY created_at ASC""",
            (session_id,),
        )

    # In-memory fallback
    return list(_messages.get(session_id, []))

File: aib/server/src/services/session.js
const db = require('../config/database');

/**
 * Create a new session
 * @returns {Promise<object>} - The created session
 */
async function createSession() {
  const result = await db.query(
    `INSERT INTO sessions (status) VALUES ('active') RETURNING *`
  );
  return result.rows[0];
}

/**
 * Get a session by ID
 * @param {string} sessionId - UUID
 * @returns {Promise<object|null>}
 */
async function getSession(sessionId) {
  const result = await db.query(
    `SELECT * FROM sessions WHERE id = $1`,
    [sessionId]
  );
  return result.rows[0] || null;
}

/**
 * Get session with its messages
 * @param {string} sessionId - UUID
 * @returns {Promise<object|null>}
 */
async function getSessionWithMessages(sessionId) {
  const session = await getSession(sessionId);
  if (!session) return null;

  const messages = await db.query(
    `SELECT id, role, content, attachments, created_at 
     FROM messages 
     WHERE session_id = $1 
     ORDER BY created_at ASC`,
    [sessionId]
  );

  return { ...session, messages: messages.rows };
}

/**
 * List sessions with optional status filter
 * @param {string} status - Optional status filter
 * @returns {Promise<Array>}
 */
async function listSessions(status) {
  let query = `SELECT * FROM sessions`;
  const params = [];

  if (status) {
    query += ` WHERE status = $1`;
    params.push(status);
  }

  query += ` ORDER BY created_at DESC`;

  const result = await db.query(query, params);
  return result.rows;
}

/**
 * Update session status
 * @param {string} sessionId - UUID
 * @param {string} status - New status
 * @returns {Promise<object>}
 */
async function updateSessionStatus(sessionId, status) {
  const updates = { status, updated_at: new Date() };
  if (status === 'completed') {
    updates.completed_at = new Date();
  }

  const result = await db.query(
    `UPDATE sessions 
     SET status = $1, updated_at = $2, completed_at = $3
     WHERE id = $4 
     RETURNING *`,
    [updates.status, updates.updated_at, updates.completed_at || null, sessionId]
  );
  return result.rows[0];
}

/**
 * Add a message to a session
 * @param {string} sessionId - UUID
 * @param {string} role - 'user' or 'assistant'
 * @param {string} content - Message text
 * @param {Array} attachments - Optional file metadata
 * @returns {Promise<object>}
 */
async function addMessage(sessionId, role, content, attachments = []) {
  const result = await db.query(
    `INSERT INTO messages (session_id, role, content, attachments) 
     VALUES ($1, $2, $3, $4) 
     RETURNING *`,
    [sessionId, role, content, JSON.stringify(attachments)]
  );

  // Update session timestamp
  await db.query(
    `UPDATE sessions SET updated_at = NOW() WHERE id = $1`,
    [sessionId]
  );

  return result.rows[0];
}

/**
 * Get all messages for a session
 * @param {string} sessionId - UUID
 * @returns {Promise<Array>}
 */
async function getMessages(sessionId) {
  const result = await db.query(
    `SELECT id, role, content, attachments, created_at 
     FROM messages 
     WHERE session_id = $1 
     ORDER BY created_at ASC`,
    [sessionId]
  );
  return result.rows;
}

module.exports = {
  createSession,
  getSession,
  getSessionWithMessages,
  listSessions,
  updateSessionStatus,
  addMessage,
  getMessages,
};
